source/lac/sparsity_pattern.cc

```diff
@@ -494,7 +494,7 @@ SparsityPattern::copy_from(const DynamicSparsityPattern &dsp)
   for (size_type i = 0; i < dsp.n_rows(); ++i)
     {
       row_lengths[i] = dsp.row_length(i);
-      if (do_diag_optimize && !dsp.exists(i, i))
+      if (do_diag_optimize && (dsp.row_length(i) == 0))
```

Comment:
This as well fixes the read. An alternative approach would be to iterate over the IndexSet instead.

Reply:
Do you have access to the index set of the DSP? If so, that might be the better way to test whether a line is locally stored.